List.fold_assoc_{left,right} + Hashtbl.fold_pair

gasche opened the pull request:
```
val List.fold_assoc_left : ('a -> 'k -> 'v -> 'a) -> 'a -> ('k * 'v) list -> 'a
val List.fold_assoc_right : ('k -> 'v -> 'a -> 'a) -> ('k * 'v) list -> 'a -> 'a
val Hashtbl.fold_pair : ('k * 'v -> 'a -> 'a) -> ('k, 'v) Hashtbl.t -> 'a -> 'a
```

These functions were born from the ashes of @nojb's #2236 (Hashtbl.{to,of}_list). The `List.fold_assoc*` can build an associative collection (`Hashtbl`, `Map`) from an associative list easily.

```
List.fold_assoc_right Hashtbl.add bindings table
```

`Hashtbl.fold_pair` can build a non-associative collection (`List`, `Stack`, `Queue`...) from a hashtable easily.

```
Hashtbl.fold_pair Queue.add tbl (Queue.create ())
```

Remark: instead of such functions, one could consider using generic currying and uncurrying combinators, as proposed in the defunct #9425. One argument in favor of the approach in the current PR is that uses of the new functions are readable in a fairly natural/accessible, even to people who are not already aware of the concept of currification.

Remark: other associative collections, in particular Map, would benefit from a `fold_pair` for the same reason. One thing at a time!

Remark: this would go in the direction of exposing two generic export functions for datastructures, a fold (here `fold_pair`), and a sequence. This corresponds to exposing a "generator" and an "iterator" in the language of [this blog post](http://gallium.inria.fr/blog/generators-iterators-control-and-continuations/) that argues that those are extremal points in the design space.

gasche committed:
List.fold_assoc_{left,right}

    val List.fold_assoc_left : ('a -> 'k -> 'v -> 'a) -> 'a -> ('k * 'v) list -> 'a
    val List.fold_assoc_right : ('k -> 'v -> 'a -> 'a) -> ('k * 'v) list -> 'a -> 'a

These functions allow to build a pure associative collection (Map)
out of an associative list easily:

    List.fold_assoc_right IntMap.add bindings IntMap.empty

gasche committed:
[minor] testsuite: refactor lib-hashtbl/htbl.ml

dbuenzli commented:
The specialisations for assoc lists seems a bit a can of worms to me (what about adding `iter_assoc_{left,right}`, etc.). I would more be in favor of adding: 
```
Hashtbl.add_pair : ('a, 'b) t -> ('a * 'b) -> unit
```
That way you can simply `List.iter Hashtbl.add_pair` an assoc which is also clearer about its imperative nature.

Regarding `Hashtbl.fold_pair` I find the name doesn't read well. Besides I'm not sure it's worth adding the cognitive burden of a entry point for the cases you mention, usually with `Hasthbl.fold` a short anonymous function will do and will stay on a single line. 

Since this was supposed to help with `Hashtbl.to_list` I would rather propose to readd this function under the name `Hashtbl.bindings` for uniformity with `Map`.

xvw commented on stdlib/list.ml:
```diff
@@ -220,6 +220,16 @@ let rec remove_assq x = function
   | [] -> []
   | (a, _ as pair) :: l -> if a == x then l else pair :: remove_assq x l
 
+let rec fold_assoc_left f accu l =
```
The question is maybe naïve but why not writting `fold_assoc_left/right` using fold_left/right?

```ocaml
let fold_assoc_left f accu l = 
    fold_left (fun accu (k, v) -> f accu k v) accu l 
```

gasche replied:
Having an inlined definition is expected to be slightly more efficient. In a stdlib, I think we can trade a little redundancy in exchange for efficiency, when the duplication is not too bad maintenance-wise. For `fold_assoc_left` I redefined as the code is simple and natural, but notice that for `Hashtbl.fold_pair` I did use the approach you mention, because the code for folding over a hashtable is sensibly more delicate, so I didn't want to duplicate it.

xvw replied:
Thanks for the reply!

gasche commented:
> The specialisations for assoc lists seems a bit a can of worms to me (what about adding `iter_assoc_{left,right}`, etc.).

I see nothing wrong with proposing `List.iter_assoc` as a counterpart to `List.assoc` on associative lists, if someone sees the need. (For `fold` I expose `fold_{left,right}` for consistency with the non-associative counterpart does, but for `iter` we just have `List.iter`.)

> I would more be in favor of adding:
> 
> ```
> Hashtbl.add_pair : ('a, 'b) t -> ('a * 'b) -> unit
> ```
> 
> That way you can simply `List.iter Hashtbl.add_pair` an assoc which is also clearer about its imperative nature.

There are two things to unpack in your comment:
1. My code example with `List.fold_assoc_right Hashtbl.add` is wrong because `Hashtbl.add` returns a unit, not the updated hashtable. So indeed if we want to allow this we need to have `List.iter_assoc` in addition. I may add it to the current PR.

2. We have a choice between proposing a curried version of the addition function, or an uncurried version of the fold. I initially proposed the approach you are suggesting now, `.add_pair`, but I think that the uncurried fold is better. The idea, as I mentioned it, is that an uncurried fold corresponds to exposing an iterator interface for the non-associative view of an associative collection (an associative collection `('foo, 'bar) t` can be seen as an non-associative collection `('foo * 'bar) nonassoc_t`. I am confident that this will be useful for many other things than just adding elements in another structure. On the other hand, I see no inherent justification in adding `.add_pair`; and to get the same flexibility we should also add `.replace_pair`, etc.; if the idea is to just propose uncurried versions of *all* key-value operations in each associative structure (or at least all those that we can expect to pass to a `fold` or `iter`), then *that* is a big can of worms indeed.


> Regarding `Hashtbl.fold_pair` I find the name doesn't read well.

Maybe `Hashtbl.fold_bindings`? (I could be persuaded to add an `.iter_bindings` as well.)

> Since this was supposed to help with `Hashtbl.to_list` I would rather propose to readd this function under the name `Hashtbl.bindings` for uniformity with `Map`.

Note that `.{fold,iter}_bindings` are much more flexible than `.bindings`, as it can be used to populate any other non-associative structure. (So does `to_seq` combined to a fold there, but note that conceptually this is less efficient, as `to_seq` gives control to the consumer, instead of the producer having control. `to_seq` would be more efficient for building datastructures that have more efficient in-control builders, such as building a balanced search tree from an ordered sequence.)

dbuenzli commented:
To be honest I find the "associative vs non-assocative" populating from a hashtable a bit architecture-astronauting. So many words... so many new entry points.  I really think `Hasthbl.{fold,iter}` is more than enough for most needs and the adaptor a one-liner if needed.

So still very unconvinced by `Hashtbl.{fold,iter}_bindings`. I would rather add the following to `List`. 

```
List.add_assoc : 'a -> 'b -> ('a, 'b) list -> ('a, 'b) list
```
that way we simply have: 
```
let to_list h = Hasthbl.fold List.add_assoc h []
```
(but would still like the `Map.bindings` uniformity)

Other than that I can be convinced by the addition of `List.iter_assoc` and `List.fold_{left,right}_assoc_`, on the ground that it provides the same traversals as we have on `Map` and `Hashtbl`.

gasche commented:
I didn't know about "architecture astronaut" as a thing; apparently it started in a [2001 blog post in Joel on Software](https://www.joelonsoftware.com/2001/04/21/dont-let-architecture-astronauts-scare-you/) that I find fairly amusing to read in retrospect -- it is harshly criticizing "hype" ideas, some of which (okay, not SOAP or XML) have since become wildly successful. If `Hashtbl.fold_bindings` does about as well as those "like Napster but missing the point" clones, I'm okay with that.

Re. `List.add_assoc`: it *is* better than `Hashtbl.add_pair`, in that it makes sense with respect to the rest of the interface, but it remains a fair amount less flexible than having uncurried fold+iter: it doesn't let you create a Stack/Queue/Rev_list of pairs easily. It might be a good compromise (it's less flexible but someone likes it more, so maybe it's more consensual?).

>  I really think Hasthbl.{fold,iter} is more than enough for most needs and the adaptor a one-liner if needed.

Well I think this is also a sensible approach: if things are not curried the way you want, write an anonymous function. Should I close the PR because all of the proposals it makes can be killed by this reasoning?

Some Stdlib PRs, such as [my partition_map that @dbuenzli forgot to say he is now happy about](https://github.com/ocaml/ocaml/pull/9066), implement things that require a real implementation effort if they are missing. It's easier to justify them.

Then there are other PRs of the form "it would be *shorter* if we could do X", and here it's very hard to set clear boundaries on what is and isn't acceptably shorter. I killed @nojb's proposal to have `Hashtbl.add_list t bindings` because I thought it would be more flexible to have people write `List.iter_assoc (Hashtbl.add t) bindings`, adding a combinator that could serve other purposes, and now @dbuenzli is proposing to write `List.iter (fun (k, v) -> Hashtbl.add t k v) bindings` instead. The argument is not transitive: each step might be convincing on its own, but clearly there *is* something annoying about writing `let t = Hashtbl.create 42 in List.iter (fun (k, v) -> Hashtbl.add t k v) bindings` instead of `Hashtbl.add_list (Hashtbl.create 42) bindings`. So what do we do now, nothing? It's the easiest solution.

dbuenzli commented:
> Re. `List.add_assoc`: it _is_ better than `Hashtbl.add_pair`, in that it makes sense with respect to the rest of the interface

I think it's better not to go the `Hasthbl.add_pair` route for the reasons you mentioned before.

Adding `List.add_assoc` makes what a proper `List.Assoc` submodule with basic commonalities of `Map.S`  would have achieved (see below).

> but it remains a fair amount less flexible than having uncurried fold+iter: it doesn't let you create a Stack/Queue/Rev_list of pairs easily.

Yes it's less flexible. But I don't understand. It's not about creating a Stack/Queue/Rev_list out of *of pairs*, it's about creating these from a *hashtable*. Except for assoc lists (which is catered by `List.add_assoc`), how many times in your programs did you do that ? 

> I killed @nojb's proposal to have Hashtbl.add_list t bindings because I thought it would be more flexible to have people write List.iter_assoc (Hashtbl.add t) bindings, adding a combinator that could serve other purposes, and now @dbuenzli is proposing to write List.iter (fun (k, v) -> Hashtbl.add t k v) bindings instead

In [this message](https://github.com/ocaml/ocaml/pull/9671#issuecomment-643613793) I said I was fine with `List.{iter,fold_left,fold_right}_assoc`. Though now thinking about it maybe it would be better to have a `List.Assoc` submodule that provides basic common functions with `Map.S`.

> The argument is not transitive: each step might be convincing on its own, but clearly there _is_ something annoying about writing `let t = Hashtbl.create 42 in List.iter (fun (k, v) -> Hashtbl.add t k v) bindings` instead of `Hashtbl.add_list (Hashtbl.create 42) bindings`. So what do we do now, nothing? It's the easiest solution.

As usual I'm not really interested in what people write, I'm interested in what people read. With respect to this I don't think its worth to multiply traversal interfaces in `Hashtbl`. At that rate, since you were then planning to go on fix this in `Map` aswell I rather have `curry` and `uncurry` which provide a general solution to "the problem" of populating non-associative structures from an associative structure and vice-versa. 

Personally I think for the initial problem at hand neither the curry combinators nor your uncurried folds are needed. All of the following would read fine to me:

```ocaml
let t = Hashtbl.create 42 in List.Assoc.iter (Hashtbl.add t) bindings 
let t = Hashtbl.create 42 in List.iter_assoc (Hasthbl.add t) bindings
let bindings = Hashtbl.fold List.Assoc.add t [] 
let bindings = Hashtbl.fold List.add_assoc t [] 
let bindings = Hashtbl.bindings t 
```


P.S. This is OT but:
> Some Stdlib PRs, such as [my partition_map that @dbuenzli forgot to say he is now happy about](https://github.com/ocaml/ocaml/pull/9066),

I'm glad to be *especially* wrong on that one. My stance on that PR is clarified by the first sentence of [that comment](https://github.com/ocaml/ocaml/pull/9066#issuecomment-607708823).

gasche commented:
I will add `List.add_assoc` and `List.iter_assoc` to this PR, and forget about `Hashtbl.add_pair` or `Hashtbl.fold_pair`.

gasche committed:
List.iter_assoc : ('k -> 'v -> unit) -> ('k * 'v) list -> unit

This function makes it easy to populate an imperative datastructure
for a list of key-value pairs:

    List.iter_assoc (Hashtbl.add tbl) bindings

gasche committed:
List.cons_assoc: 'k -> 'v -> ('k * 'v) list -> ('k * 'v) list

gasche commented:
I pushed `iter_assoc` and `cons_assoc` (I think that `List.cons_assoc` is more consistent with `List.cons` than `List.add_assoc` would be, but I can be persuaded otherwise.)

I've been mellowing out on the idea of having an `Assoc` submodule (provided we don't deprecate existing `*_assoc` functions, I hate dealing with deprecation). Should we go the `List.Assoc` route? Other opinions would help here.

```
module Assoc : sig
  type ('k, 'v) t = ('k * 'v) list

  val empty : ('k, 'v) t
  val cons : 'k -> 'v -> ('k, 'v) t -> ('k, 'v) t

  val find : 'k -> ('k, 'v) t -> 't option
  val findq : 'k -> ('k, 'v) t -> 't option

  val mem : 'k -> ('k, 'v) t -> bool
  val memq : 'k -> ('k, 'v) t -> bool

  val remove : 'k -> ('k, 'v) t -> ('k, 'v) t
  val removeq : 'k -> ('k, 'v) t -> ('k, 'v) t

  val map : ('k -> 'v1 -> 'v2) -> ('k, 'v1) t -> ('k, 'v2) t
  val iter : ('k -> 'v -> unit) -> ('k, 'v) t -> unit
  val fold_{left,right} : ...
end
```

... or we could be more daring and have


```
module Assoc : sig
  type ('k, 'v) t = ('k * 'v) list

  val empty : ('k, 'v) t
  val cons : 'k -> 'v -> ('k, 'v) t -> ('k, 'v) t

  val find : 'k -> ('k, 'v) t -> 't option
  val find_by : ('k -> 'k -> bool) -> 'k -> ('k, 'v) t -> 't option

  val mem : 'k -> ('k, 'v) t -> bool
  val mem_by : ('k -> 'k -> bool) -> 'k -> ('k, 'v) t -> bool

  val remove : 'k -> ('k, 'v) t -> ('k, 'v) t
  val remove_by : ('k -> 'k -> bool) -> 'k -> ('k, 'v) t -> ('k, 'v) t

  val map : ('k -> 'v1 -> 'v2) -> ('k, 'v1) t -> ('k, 'v2) t
  val iter : ('k -> 'v -> unit) -> ('k, 'v) t -> unit
  val fold_{left,right} : ...
end
```

yallop commented:
Another (perhaps even more daring) possibility, which would avoid having different versions of various functions for different equalities, is to add a functorized interface similar to `Map.S`:

```ocaml
module type Assoc = sig
  type k
  type 'v t

  val empty : 'v t
  val cons : k -> 'v -> 'v t -> 'v t
  val find : k -> 'v t -> 't option
  val mem : k -> 'v t -> bool
  val remove : k -> 'v t -> 'v t
  (* ... *)
end
```

gasche commented:
@yallop: right now people use associative maps as a "quick fix" solution to dictionaries; the functorized interface makes them as heavy to use as Map, which I see as a losing proposition as it is generally more reasonable to use Map in this case. (Well, unless you have only equality and no ordering on your key data-type, but I never personally encountered this situation and I'm not sure it is worth supporting in the stdlib, especially given the potential for mis-use to write slow programs.)

dbuenzli commented:
> (I think that `List.cons_assoc` is more consistent with `List.cons` than `List.add_assoc` would be, but I can be persuaded otherwise.)

I would rather argue for consistency with a key-value interface than with `List`. If you have  `List.Assoc.add` you get a minimal set of functions names that are easy to remember and work (if only slightly differently) across `Assoc`, `Hashtbl` and `Map`:  `M.{mem,add,remove,find,fold,iter}`.

gasche commented:
Fair enough.

Regarding `memq` vs. `mem_by` vs. functorized interface, one simpler option would be to offer none of them and just `mem`. (We *could* have `List.Assoq`, but I'm not sure it is worth it.)
